File: board_dsl/src/main/java/com/dooho/board/service/BoardService.java
package com.dooho.board.service;

import com.dooho.board.dto.ResponseDto;
import com.dooho.board.dto.board.PatchBoardDto;
import com.dooho.board.dto.board.PatchBoardResponseDto;
import com.dooho.board.entity.BoardEntity;
import com.dooho.board.repository.BoardRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.web.multipart.MultipartFile;

import java.time.LocalDate;
import java.time.ZoneId;
import java.time.ZonedDateTime;
import java.time.format.DateTimeFormatter;
import java.util.ArrayList;
import java.util.List;

@Service
public class BoardService {

    private final BoardRepository boardRepository;
    private final FileService fileService;


    @Autowired
    public BoardService(BoardRepository boardRepository, FileService fileService) {
        this.boardRepository = boardRepository;
        this.fileService = fileService;
    }


    public ResponseDto<BoardEntity> getBoard(Integer boardNumber){
        BoardEntity board = null;

        try{
            board = boardRepository.findByBoardNumber(boardNumber);
        }catch (Exception e){
            return ResponseDto.setFailed("DataBase Error!");
        }

        return ResponseDto.setSuccess("Success",board);
    }

    public ResponseDto<BoardEntity> register(
            String boardTitle,
            String boardContent,
            String boardWriterEmail,
            String boardWriterProfile,
            String boardWriterNickname,
            String boardWriteDate,
            MultipartFile boardImage,
            MultipartFile boardVideo,
            MultipartFile boardFile){        DateTimeFormatter formatter = DateTimeFormatter.ISO_INSTANT;
        ZonedDateTime zonedDateTime = ZonedDateTime.parse(boardWriteDate, formatter.withZone(ZoneId.of("UTC")));

        LocalDate localDate =  zonedDateTime.toLocalDate();

        if(boardRepository.existsByBoardTitle(boardTitle)){
            return ResponseDto.setFailed("Same Title already exist!");
        }

        BoardEntity boardEntity = new BoardEntity();
        boardEntity.setBoardTitle(boardTitle);
        boardEntity.setBoardContent(boardContent);
        boardEntity.setBoardWriterEmail(boardWriterEmail);
        boardEntity.setBoardWriterProfile(boardWriterProfile);
        boardEntity.setBoardWriterNickname(boardWriterNickname);
        boardEntity.setBoardWriteDate(localDate);
        boardRepository.save(boardEntity);


        try{
            fileService.uploadFile(boardImage,boardVideo,boardFile,boardEntity);
            boardRepository.save(boardEntity);
        }catch (Exception e){
            return ResponseDto.setFailed("DataBase Error!");

        }

        return ResponseDto.setSuccess("Register Success!",boardEntity);
    }

    public ResponseDto<List<BoardEntity>> getTop3(){
        List<BoardEntity> boardList = new ArrayList<BoardEntity>();
        LocalDate date = LocalDate.now().minusDays(365);
        try{
            boardList = boardRepository.findTop3ByBoardWriteDateAfterOrderByBoardLikeCountDesc(date);

        }catch (Exception e){
            e.printStackTrace();
            return ResponseDto.setFailed("DataBase Error");
        }

        return ResponseDto.setSuccess("Success",boardList);
    }

    public ResponseDto<List<BoardEntity>> getList(){
        List<BoardEntity> boardList = new ArrayList<>();

        try{
            boardList = boardRepository.findByOrderByBoardWriteDateDescBoardNumberDesc();
        }catch (Exception e){
            e.printStackTrace();
            return ResponseDto.setFailed("DataBase Error");
        }

        return ResponseDto.setSuccess("Success", boardList);
    }


    public ResponseDto<?> deleteBoard(Integer boardNumber) {

        try{
            System.out.println("boardNumber = " + boardNumber);
            boardRepository.deleteBoardEntityByBoardNumber(boardNumber);
        }catch (Exception e){
            e.printStackTrace();
            return ResponseDto.setFailed("DataBase Error!");
        }
        return ResponseDto.setSuccess("Success",null);
    }

    public ResponseDto<?> increaseView(Integer boardNumber, Integer increase) {
        BoardEntity boardEntity = boardRepository.findByBoardNumber(boardNumber);
        Integer boardClick = boardEntity.getBoardClickCount();
        try{
            boardEntity.setBoardClickCount(boardClick + increase);
            boardRepository.save(boardEntity);
        }catch (Exception e){
            e.printStackTrace();
            return ResponseDto.setFailed("DataBase Error!");
        }
        return ResponseDto.setSuccess("Success",null);
    }

    public ResponseDto<PatchBoardResponseDto> editBoard(Integer boardNumber, PatchBoardDto dto) {
        BoardEntity board = null;
        String boardTitle = dto.getBoardTitle();
        String boardContent = dto.getBoardContent();
        LocalDate boardWriteDate = dto.getBoardWriteDate();
        try{
            board = boardRepository.findByBoardNumber(boardNumber);
            board.setBoardTitle(boardTitle);
            board.setBoardContent(boardContent);
            board.setBoardWriteDate(boardWriteDate);

            boardRepository.save(board);

        }catch (Exception e){
            e.printStackTrace();
            return ResponseDto.setFailed("DataBase Error!");
        }

        PatchBoardResponseDto patchBoardResponseDto = new PatchBoardResponseDto(board);

        return ResponseDto.setSuccess("Success!",patchBoardResponseDto);
    }
}

File: board_spring/src/main/java/com/dooho/board/api/board/BoardService.kt
package com.dooho.board.api.board

import com.dooho.board.api.ResponseDto
import com.dooho.board.api.ResponseDto.Companion.setFailed
import com.dooho.board.api.ResponseDto.Companion.setSuccess
import com.dooho.board.api.file.FileService
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service
import org.springframework.web.multipart.MultipartFile
import java.time.LocalDate
import java.time.ZoneId
import java.time.ZonedDateTime
import java.time.format.DateTimeFormatter

@Service
class BoardService @Autowired constructor(
    private val boardRepository: BoardRepository,
    private val fileService: FileService
) {
    fun getBoard(boardNumber: Int?): ResponseDto<BoardEntity?> {
        var board: BoardEntity? = null
        board = try {
            boardRepository.findByBoardNumber(boardNumber)
        } catch (e: Exception) {
            return setFailed("Database Error!")
        }
        return setSuccess("Success", board)
    }


    fun register(
        boardTitle: String?,
        boardContent: String?,
        boardWriterEmail: String?,
        boardWriterProfile: String?,
        boardWriterNickname: String?,
        boardWriteDate: String?,
        boardImage: MultipartFile?,
        boardVideo: MultipartFile?,
        boardFile: MultipartFile?
    ): ResponseDto<BoardEntity?> {
        val formatter = DateTimeFormatter.ISO_INSTANT
        val zonedDateTime = ZonedDateTime.parse(boardWriteDate, formatter.withZone(ZoneId.of("UTC")))
        val localDate = zonedDateTime.toLocalDate()
        val boardEntity = BoardEntity(
            boardTitle = boardTitle,
            boardContent = boardContent,
            boardWriterEmail = boardWriterEmail,
            boardWriterProfile = boardWriterProfile,
            boardWriterNickname = boardWriterNickname,
            boardWriteDate = localDate
        )

        if (boardRepository.existsByBoardTitle(boardTitle)) {
            return setFailed("Same Title already exists!")
        }

        boardRepository.save(boardEntity)

        try {
            fileService.uploadFile(boardImage, boardVideo, boardFile, boardEntity)
            boardRepository.save(boardEntity)
        } catch (e: Exception) {
            return setFailed("Database Error!")
        }

        return setSuccess("Register Success!", boardEntity)
    }
    fun getTop3() : ResponseDto<List<BoardEntity?>?>{
        var boardList: List<BoardEntity?>? = ArrayList()
        var date = LocalDate.now().minusDays(365)
        try{
            boardList = boardRepository.findTop3ByBoardWriteDateAfterOrderByBoardLikeCountDesc(date)
        }catch (e : Exception){
            e.printStackTrace()
            return setFailed("DataBase Error")
        }
        return setSuccess("Success",boardList)
    }

/**    코틀린에선, 위 메소드를 아래와 같이 프로퍼티 형태로 적용 가능
 * val top3: ResponseDto<List<BoardEntity?>?>
        get() {
            var boardList: List<BoardEntity?>? = ArrayList()
            val date = LocalDate.now().minusDays(365)
            boardList = try {
                boardRepository.findTop3ByBoardWriteDateAfterOrderByBoardLikeCountDesc(date)
            } catch (e: Exception) {
                e.printStackTrace()
                return setFailed<List<BoardEntity?>?>("DataBase Error")
            }
            return setSuccess("Success", boardList)
        }
**/

    fun getList(): ResponseDto<List<BoardEntity?>?>{
        var boardList : List<BoardEntity?>? = ArrayList()
        try{
            boardList = boardRepository.findByOrderByBoardWriteDateDescBoardNumberDesc()
        }catch (e : Exception){
            e.printStackTrace()
            return setFailed("DataBase Error")
        }
        return setSuccess("Success",boardList)
    }

    fun deleteBoard(boardNumber: Int?): ResponseDto<String?> {
        try {
            println("boardNumber = $boardNumber")
            boardRepository.deleteBoardEntityByBoardNumber(boardNumber)
        } catch (e: Exception) {
            e.printStackTrace()
            return setFailed("DataBase Error!")
        }
        return setSuccess("Success", "Success")
    }


    fun editBoard(boardNumber: Int?, dto: PatchBoardDto?): ResponseDto<PatchBoardResponseDto?> {
        var board: BoardEntity? = null
        val boardTitle = dto?.boardTitle
        val boardContent = dto?.boardContent
        val boardWriteDate = dto?.boardWriteDate
        try {
            board = boardRepository.findByBoardNumber(boardNumber)
            board?.boardTitle = boardTitle
            board?.boardContent = boardContent
            board?.boardWriteDate =boardWriteDate
            boardRepository.save(board)
        } catch (e: Exception) {
            e.printStackTrace()
            return setFailed("DataBase Error!")
        }
        val patchBoardResponseDto = PatchBoardResponseDto(board)
        return setSuccess("Success!", patchBoardResponseDto)
    }

    fun increaseView(boardNumber: Int?): ResponseDto<String?> {
        var board : BoardEntity? = boardRepository.findByBoardNumber(boardNumber)
        var boardClick : Int? = board?.boardClickCount
        try{
            board?.boardClickCount = boardClick?.plus(1)
            boardRepository.save(board)
        }catch(e : Exception){
            e.printStackTrace()
            return setFailed("DataBase Error!")
        }
        return setSuccess("Success", "")
    }

}